react/features/base/avatar/components/AbstractAvatar.js

```diff
@@ -0,0 +1,215 @@
+     */
+    _avatarURL: ?string,
+
+    /**
+     * URI of the gravatar if identifyable.
+     */
+    _gravatarURL: ?string,
+
+    /**
+     * The string we base the initials on (this is generated from a list of precendences).
+     */
+    _initialsBase: ?string,
+
+    /**
+     * A string to override the initials to generate a color of. This is handy if you don't want to make
+     * the background color match the string that the initials are generated from.
+     */
+    colorBase?: string,
+
+    /**
+     * Display name of the entity to render an avatar for (if any). This is handy when we need
+     * an avatar for a non-participasnt entity (e.g. a recent list item).
+     */
+    displayName?: string,
+
+    /**
+     * The ID of the participant to render an avatar for (if it's a participant avatar).
+     */
+    participantId?: string,
+
+    /**
+     * The size of the avatar.
+     */
+    size: number,
+
+    /**
+     * URI of the avatar, if any.
+     */
+    uri: ?string,
+}
+
+type State = {
+    avatarFailed: boolean,
+    gravatarFailed: boolean
+}
+
+export const DEFAULT_SIZE = 65;
+
+/**
+ * Implements an abstract class to render avatars in the app.
+ */
+export default class AbstractAvatar<P: Props> extends PureComponent<P, State> {
+    /**
+     * Instantiates a new {@code Component}.
+     *
+     * @inheritdoc
+     */
+    constructor(props: P) {
+        super(props);
+
+        this.state = {
+            avatarFailed: false,
+            gravatarFailed: false
+        };
+
+        this._onAvatarLoadError = this._onAvatarLoadError.bind(this);
+        this._onGravatarLoadError = this._onGravatarLoadError.bind(this);
+    }
+
+    /**
+     * Implements {@code Component#componentDidUpdate}.
+     *
+     * @inheritdoc
+     */
+    componentDidUpdate(prevProps: P) {
+        if (prevProps._avatarURL !== this.props._avatarURL
+            || prevProps._gravatarURL !== this.props._gravatarURL) {
+
+            // URLS changed, so we need to try to fetch them again.
+            // Eslint doesn't like this statement, but based on the React doc, it's safe if it's
+            // wrapped in a condition: https://reactjs.org/docs/react-component.html#componentdidupdate
+
+            // eslint-disable-next-line react/no-did-update-set-state
+            this.setState({
+                avatarFailed: false,
+                gravatarFailed: false
+            });
+        }
+    }
+
+    /**
+     * Implements {@code Componenr#render}.
+     *
+     * @inheritdoc
+     */
+    render() {
+        const {
+            _avatarURL,
+            _gravatarURL,
+            _initialsBase,
+            colorBase
+        } = this.props;
+        const {
+            avatarFailed,
+            gravatarFailed
+        } = this.state;
+
+        const _initials = getInitials(_initialsBase);
+
+        // Precendence logic of avatars:
+
+        if (_avatarURL && _avatarURL.match(/^http[s]*:\/\//i) && !avatarFailed) {
```

Comment:
This logic is reversed to how mobile avatars behave currently, which is IMHO the expectation: we first show the locally generaterd avatar, then, if / when the remote one is loaded we show that one instead.

Using the network conditioner to make your connection slow. Then we wouldn't show anything until the load attempt has succeeded or failed.

Reply:
Now this was refactored, thanks.

There is a caveat though: if there is an URI provided to the avatar component, we still try to load that first, before rendering any default (or initials based) avatar. In this case we don't render anything until the uri deems to be loaded or fails. The reason for this is that this is a corner case (e.g. in case of IncomingCallPage) and since the component is the one that decides if the URI can be loaded or not, we need to render the avatar (empty at first) and let them succeed or fail. Then if we render an initials based avatar, there is a short period (blink) when the avatar already loaded and rendered and the initials based avatar is still shown, so users see a duplicated avatar. I think that is a worse UX than not rendering anything for a while. But again, this is a really rare corner case.